internal/admin/admin.go

```diff
@@ -544,35 +546,50 @@ func (adm *Admin) updateLogStream(ctx context.Context, lsid types.LogStreamID, p
 
 	clusmeta, err := adm.mrmgr.ClusterMetadataView().ClusterMetadata(ctx)
 	if err != nil {
-		return nil, err
+		return nil, status.Errorf(codes.Unavailable, "update log stream: %s", err.Error())
 	}
 
 	oldLSDesc, err := clusmeta.MustHaveLogStream(lsid)
-	if err != nil {
-		return nil, err
+	if err != nil || oldLSDesc.Status.Deleted() {
+		return nil, status.Errorf(codes.NotFound, "update log stream: no such log stream %d", lsid)
 	}
 
-	status := oldLSDesc.GetStatus()
-	if status.Running() || status.Deleted() {
-		return nil, errors.Errorf("invalid log stream status: %s", status)
+	if oldLSDesc.Status.Running() {
+		return nil, status.Errorf(codes.FailedPrecondition, "update log stream: invalid log stream status %s", oldLSDesc.Status)
 	}
-
-	replace := false
-	newLSDesc := proto.Clone(oldLSDesc).(*varlogpb.LogStreamDescriptor)
-	for i := range newLSDesc.Replicas {
-		// TODO - fix? poppedReplica can ignore path.
-		if newLSDesc.Replicas[i].GetStorageNodeID() == poppedReplica.GetStorageNodeID() {
-			newLSDesc.Replicas[i] = &pushedReplica
-			replace = true
-			break
+	popIdx, pushIdx := -1, -1
+	for idx := range oldLSDesc.Replicas {
+		snid := oldLSDesc.Replicas[idx].StorageNodeID
+		if snid == poppedReplica.StorageNodeID {
+			popIdx = idx
+		} else if snid == pushedReplica.StorageNodeID {
+			pushIdx = idx
 		}
 	}
-	if !replace {
-		adm.logger.Panic("logstream push/pop error")
+	if popIdx < 0 && pushIdx >= 0 { // already updated
+		return oldLSDesc, nil
+	}
+	if popIdx < 0 && pushIdx < 0 {
+		return nil, status.Errorf(
+			codes.FailedPrecondition,
+			"update log stream: no victim replica (snid=%v) in log stream %+v",
+			poppedReplica.StorageNodeID,
+			oldLSDesc.Replicas,
+		)
+	}
+	if popIdx >= 0 && pushIdx >= 0 {
+		return nil, status.Errorf(
+			codes.FailedPrecondition,
+			"update log stream: victim replica and new replica already exist in log stream %+v",
+			oldLSDesc.Replicas,
+		)
 	}
 
+	newLSDesc := proto.Clone(oldLSDesc).(*varlogpb.LogStreamDescriptor)
+	newLSDesc.Replicas[popIdx] = &pushedReplica
+
 	if err := adm.snmgr.AddLogStreamReplica(ctx, pushedReplica.GetStorageNodeID(), newLSDesc.TopicID, lsid, pushedReplica.GetPath()); err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "update log stream")
```

Comment:
What does this error mean? How should kovadm handle it?
If we can't create LS, should we decide that we can't do UpdateLS?

Reply:
Errors returned from the storage node and metadata repository do not have codes yet. So, their codes will be `codes.Unknown`. It will be defined soon.
The client can retry UpdateLogStream when it receives an error with `codes.Unknown`. It can be resolved with retries but cannot be guaranteed bounded times. It is also okay to stop calling UpdateLogStream with the same parameters.

Reply:
Okay. We will continue to retry with the same argument. If it is a request that can never be executed, we expect it to return an appropriate error.